lib/CL/pocl_llvm_wg.cc

```diff
@@ -108,91 +121,346 @@ static llvm::TargetOptions GetTargetOptions() {
+  PB.registerCGSCCAnalyses(CGAM);
+  PB.registerFunctionAnalyses(FAM);
+  PB.registerLoopAnalyses(LAM);
+  PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
 
-  if (Machine)
-    Passes->add(
-        createTargetTransformInfoWrapperPass(Machine->getTargetIRAnalysis()));
+  pocl::registerPassBuilderPasses(PB);
 
+#ifndef SEPARATE_OPTIMIZATION_FROM_POCL_PASSES
+  OptimizationLevel Opt;
+  if (SizeLevel > 0)
+    PoclPipeline += ",default<Os>";
+  else {
+    switch (OptimizeLevel) {
+    case 0:
+      PoclPipeline += ",default<O0>";
+      break;
+    case 1:
+      PoclPipeline += ",default<O1>";
+      break;
+    case 2:
+      PoclPipeline += ",default<O2>";
+      break;
+    case 3:
+    default:
+      PoclPipeline += ",default<O3>";
+      break;
+    }
+  }
+#endif
 
-  /* Disables automated generation of libcalls from code patterns.
-     TCE doesn't have a runtime linker which could link the libs later on.
-     Also the libcalls might be harmful for WG autovectorization where we
-     want to try to vectorize the code it converts to e.g. a memset or
-     a memcpy */
-  TargetLibraryInfoImpl TLII(triple);
-  TLII.disableAllFunctions();
-  Passes->add(new TargetLibraryInfoWrapperPass(TLII));
+  return PB.parsePassPipeline(PM, StringRef(PoclPipeline));
+}
+
+void PoCLModulePassManager::run(llvm::Module &Bitcode) {
+  PM.run(Bitcode, MAM);
+#ifdef SEPARATE_OPTIMIZATION_FROM_POCL_PASSES
+  populateModulePM(nullptr, (void *)&Bitcode, OptimizeLevel, SizeLevel,
+                   Vectorize);
+#endif
+}
+
+/**
+ * @brief The TwoStagePoCLModulePassManager class for running the full pipeline
+ *
+ * the full pipeline of PoCL LLVM passes looks like this:
+ *  <pocl passes, LLVM optimization, pocl passes, LLVM optimization>
+ *
+ * this is how it worked with the old PM; however, running
+ * all of these with a single PassManager (with new PM) leads
+ * to crashes and/or failing tests. This class splits the execution of
+ * the pipeline into two halfs (stages) so both contain some pocl-passes plus
+ * the LLVM optimization passes, and uses a separate PassManager
+ * instance for both (see class PoCLModulePassManager).
+ * @note see also comment for SEPARATE_OPTIMIZATION_FROM_POCL_PASSES macro,
+ * which (if enabled) further separates optimization pipeline into its own PM
+ */
+class TwoStagePoCLModulePassManager {
+  PoCLModulePassManager Stage1;
+  PoCLModulePassManager Stage2;
+#ifndef PER_STAGE_TARGET_MACHINE
+  std::unique_ptr<llvm::TargetMachine> Machine;
+#endif
+public:
+  TwoStagePoCLModulePassManager() = default;
+  llvm::Error build(cl_device_id Dev, const std::string &S1_Pipeline,
+                    unsigned S1_OLevel, unsigned S1_SLevel,
+                    const std::string &S2_Pipeline,
+                    unsigned S2_OLevel, unsigned S2_SLevel);
+  void run(llvm::Module &Bitcode);
+};
+
+llvm::Error TwoStagePoCLModulePassManager::build(
+    cl_device_id Dev, const std::string &S1_Pipeline, unsigned S1_OLevel,
```

Comment:
```suggestion
    cl_device_id Dev, const std::string &Stage1Pipeline, unsigned Stage1OLevel,
```
Etc. to avoid the underscore.